New module 'set_fact' to define host facts

dagwieers opened the pull request:
This module allows you to set host facts (or export play variables to the playbook scope if you fancy that).

The module also accepts complex arguments.

``` yaml
 - action: set_fact fact="something" global_fact="${local_var}"'
 - action: set_fact
   args:
      fact: something
      global_fact: ${local_var}
```

steverweber commented:
How would the global fit into this priority?
http://ansible.cc/docs/playbooks2.html#understanding-variable-precedence

I would love to see something like in the list as 8. that is defined inside a play like vars.

dagwieers commented:
@steverweber: Global variables have the same priority as (and are implemented as) host facts, that's why the documentation refers to them as host facts as well. Local variables have the same priority as playbook variables (I think this is an error in the documentation as it should be called "play variables"). So local variables have priority over global variables (no matter what the order is). We could change the behavior so that global variables also set local variables, but it's up to @mpdehaan to decide how he would like the behavior.

dagwieers committed:
New module 'set_fact' to define host facts

This module allows you to set host facts (or export play variables to the playbook scope if you fancy that).

The module also accepts complex arguments.

```yaml
 - action: set_fact fact="something" global_fact="${local_var}"'
 - action: set_fact
   args:
      fact: something
      global_fact: ${local_var}
```

mpdehaan commented:
So technically it's not global, it's setting a variable for all hosts in the current play that have the task.

This feels like this does "set" to me, but does not set a global.

For instance, you could pass in a host specific variable.

dagwieers commented:
It's global in the sense that the variable is know beyond the scope of the play, whereas local means in the scope of the play. This is exactly what both set and global do at the moment.

Of course it's not global, in the sense that it is one variable for all hosts, that's not the intention because the nice thing is that it allows to set variables conditionally (per host).

mpdehaan commented:
Yeah, so naming is important here, unfortunately, since we can't call it
global.

Go back to "set"  ?

On Wed, Apr 17, 2013 at 12:07 PM, Dag Wieers notifications@github.comwrote:

> It's global in the sense that the variable is know beyond the scope of the
> play, whereas local means in the scope of the play. This is exactly what
> both set and global do at the moment.
> 
> Of course it's not global, in the sense that it is one variable for all
> hosts, that's not the intention because the nice thing is that it allows to
> set variables conditionally (per host).
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/ansible/ansible/pull/2688#issuecomment-16515257
> .

dagwieers commented:
But that makes no sense, set and local are almost alike. Nothing here indicates to what scope it is, only local/global clearly say that there is a scope. set/local/define/var it's all interchangeable in this case.

Before we continu with this discussion, what functionality do you actually want/need. I made that clear with the export module:
- Set a variable for the playbook scope, so that you can either export a local (play) variable or define something new conditionally.

My initial mail clearly called it "Exporting variables as host facts" and the module was called "export". Why someone wants to set local vars is beyond me, but that's what "vars:" and "vars_files:" already do which I think is limited in use.

Maybe this module should just be called "fact", and we drop the local (play scope) module altogether. At least "fact" is very clearly a host-specific property, whereas variable IMO could mean anything. I think we lack the proper nomenclature which is confusing...

mpdehaan commented:
So local would be something like like vars_files that was true and exactly the same for all hosts.

We've had a lot of interest at being able to do things like:
- tasks:
  - set_local: x={{ asdf \* 10 }}

Which is I thought what this was trying to enable, where asdf may vary between hosts.

If so, the idea and implementation is fine, it's just not global by any means that a developer would expect it to be.   Just a matter of naming.

Facts are already a thing, this feels entirely like it is a "set_local" to me and it's named (no disrespect intended) the opposite of what it is actually doing.

dagwieers commented:
global for you means 'for all hosts', while for me it means 'for all plays'.
local for you means 'for one host', while for me it means 'for one play'.

Correct ?

So I would ditch the local/global nomenclature, and simply go back to what this module does, it creates facts. Facts by nature are host-specific and for all plays, so I really think we should call it "fact". Since we use the ansible_facts mechanism we aren't even lying and for the precedence rules it's also very clear.

bcoca commented:
set_fact and set_var make sense?

Brian Coca

mpdehaan commented:
set_fact and set_var does not make sense to me.

One is inventory scope, one is larger (global) scope.

On Wed, Apr 17, 2013 at 8:30 PM, Brian Coca notifications@github.comwrote:

> set_fact and set_var make sense?
> 
> Brian Coca
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/ansible/ansible/pull/2688#issuecomment-16544360
> .

mpdehaan commented:
sorry, just the way my brain works I guess?

Both are variables, so calling one "var" doesn't work for me.

Names are hard!  :)

On Wed, Apr 17, 2013 at 11:17 PM, Michael DeHaan
michael.dehaan@gmail.comwrote:

> set_fact and set_var does not make sense to me.
> 
> One is inventory scope, one is larger (global) scope.
> 
> On Wed, Apr 17, 2013 at 8:30 PM, Brian Coca notifications@github.comwrote:
> 
> > set_fact and set_var make sense?
> > 
> > Brian Coca
> > 
> > —
> > Reply to this email directly or view it on GitHubhttps://github.com/ansible/ansible/pull/2688#issuecomment-16544360
> > .

bcoca commented:
set_hostvar (the global/per host across plays)
set_var (local/specific to play, works like var: )

dagwieers commented:
If it looks like a fact, and it acts like a fact, and it is implement as ansible_facts, _it is a fact_. Why don't we just call it a fact ?

http://ansible.cc/docs/playbooks2.html#understanding-variable-precedence

I don't understand the tendency to call everything a variable. I'd simply call the module "fact", it's clear, concise, makes precedence obvious and leaves no room for interpretation. And it's **not** in the inventory scope, it's in the facts scope.

mpdehaan commented:
facts feel like inventory scope to me, being there are only two scopes,
inventory scope and not inventory scope, meaning $foo can vary by host with
facts.

I actually like Brian's last suggestion, but I'm increasingly feeling
set_fact is ok too.

(Though the other set_var thing, I don't know, is there a need for that?)

On Thu, Apr 18, 2013 at 3:59 PM, Dag Wieers notifications@github.comwrote:

> If it looks like a fact, and it acts like a fact, and it is implement as
> ansible_facts, _it is a fact_. Why don't we just call it a fact ?
> 
> http://ansible.cc/docs/playbooks2.html#understanding-variable-precedence
> 
> I don't understand the tendency to call everything a variable. I'd simply
> call the module "fact", it's clear, concise, makes precedence obvious and
> leaves no room for interpretation. And it's _not_ in the inventory scope,
> it's in the facts scope.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/ansible/ansible/pull/2688#issuecomment-16603859
> .

dagwieers commented:
Michael, what does it mean "feels like inventory scope" ? Looking at the variable precedence documentation, inventory and facts are **not** the same. And this module implements facts.
1. Any variables specified with –extra-vars (-e) on the ansible-playbook command line.
2. Variables loaded from YAML files mentioned in ‘vars_files’ in a playbook.
3. **facts, whether built in or custom, or variables assigned from the ‘register’ keyword.**
4. variables passed to parameterized task include statements.
5. ‘vars’ as defined in the playbook.
6. **Host variables from inventory.**
7. Group variables from inventory in inheritance order. This means if a group includes a sub-group, the variables in - the subgroup have higher precedence.

dagwieers commented:
So what I am basically saying is that if we have so many rules for various kinds of variables, we should start to have a proper nomenclature so everybody understand what we are talking about. Mixing facts and host inventory variables is a good way to add confusion, where it is important to know what something is in order to understand what is going to happen.

mpdehaan commented:
I just mean that, in a given host loop, the value of "$foo" is different
between hosts for facts and inventory variables -- I'm calling this
inventory scope.

(So I'm agreeing with you, I think in the end, that set_fact is a legit
name)

So this may come from a slight confusion between the difference between
"variable precedence order" (what overrides what) and scope, which is a
slightly different animal :)

I am ok with set_fact for this one, it's definitely not global as the value
is different for different host objects depending on how used, but set_fact
is ok.

On Thu, Apr 18, 2013 at 4:08 PM, Dag Wieers notifications@github.comwrote:

> Michael, what does it mean "feels like inventory scope" ? Looking at the
> variable precedence documentation, inventory and facts are _not_ the
> same. And this module implements facts.
> 1. Any variables specified with –extra-vars (-e) on the
>    ansible-playbook command line.
> 2. Variables loaded from YAML files mentioned in ‘vars_files’ in a
>    playbook.
> 3. _facts, whether built in or custom, or variables assigned from the
>    ‘register’ keyword._
> 4. variables passed to parameterized task include statements.
> 5. ‘vars’ as defined in the playbook.
> 6. _Host variables from inventory._
> 7. Group variables from inventory in inheritance order. This means if
>    a group includes a sub-group, the variables in - the subgroup have higher
>    precedence.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/ansible/ansible/pull/2688#issuecomment-16604341
> .

mpdehaan commented:
I tend to view facts and host inventory variables as ultimately thigns that
intermingle in the same scope and namespace.

also, Words are hard!

On Thu, Apr 18, 2013 at 4:11 PM, Dag Wieers notifications@github.comwrote:

> So what I am basically saying is that if we have so many rules for various
> kinds of variables, we should start to have a proper nomenclature so
> everybody understand what we are talking about. Mixing facts and host
> inventory variables is a good way to add confusion, where it is important
> to know what something is in order to understand what is going to happen.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/ansible/ansible/pull/2688#issuecomment-16604471
> .

dagwieers commented:
set_fact it will become, the master has ruled ;-)

mpdehaan commented:
Great ansible conclave of 2013 has chosen a module!

Unleash the white smoke!

On Thu, Apr 18, 2013 at 4:18 PM, Dag Wieers notifications@github.comwrote:

> set_fact it will become, the master has ruled ;-)
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/ansible/ansible/pull/2688#issuecomment-16604964
> .

dagwieers commented:
:-)

bcoca commented:
finally!

On Thu, Apr 18, 2013 at 4:27 PM, Michael DeHaan notifications@github.comwrote:

> Great ansible conclave of 2013 has chosen a module!
> 
> Unleash the white smoke!
> 
> On Thu, Apr 18, 2013 at 4:18 PM, Dag Wieers notifications@github.comwrote:
> 
> > set_fact it will become, the master has ruled ;-)
> > 
> > —
> > Reply to this email directly or view it on GitHub<
> > https://github.com/ansible/ansible/pull/2688#issuecomment-16604964>
> > .
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/ansible/ansible/pull/2688#issuecomment-16605616
> .

## 

Brian Coca
Stultorum infinitus est numerus
0110000101110010011001010110111000100111011101000010000001111001011011110111010100100000011100110110110101100001011100100111010000100001
Pedo mellon a minno

mpdehaan commented:
merged, thanks!
